File: dotnet/Trinsic/ServiceBase.cs
using System.Diagnostics.CodeAnalysis;
using System.Reflection;
using System.Runtime.CompilerServices;
using Grpc.Core;
using Google.Protobuf;
using Trinsic.Services.Common.V1;
using Trinsic.Services.Account.V1;
using System.Threading.Tasks;
using Grpc.Net.Client;
using Trinsic.Sdk.Options.V1;
#if __BROWSER__
using System.Net.Http;
using Grpc.Net.Client.Web;
#endif

namespace Trinsic;

[SuppressMessage("Usage", "CA2211:Non-constant fields should not be visible")]
public abstract class ServiceBase
{
    internal const string DefaultEcosystem = "";
    internal const bool DefaultServerUseTls = true;
    internal const int DefaultServerPort = 443;
    internal const string DefaultServerEndpoint = "prod.trinsic.cloud";

    protected internal ServiceBase() : this(new()) { }

    protected internal ServiceBase(ServiceOptions options) {
        Options = options;
        EnsureOptionDefaults();
        Channel = CreateChannel(Options);
    }

    private void EnsureOptionDefaults() {
        if (string.IsNullOrWhiteSpace(Options.ServerEndpoint)) Options.ServerEndpoint = DefaultServerEndpoint;
        if (Options.ServerPort == default) Options.ServerPort = DefaultServerPort;
        if (Options.ServerPort == DefaultServerPort) Options.ServerUseTls = DefaultServerUseTls;
    }

    private static GrpcChannel CreateChannel(ServiceOptions options) {
#if __BROWSER__
        var httpClient = new HttpClient(new GrpcWebHandler(GrpcWebMode.GrpcWeb, new HttpClientHandler()));
        return GrpcChannel.ForAddress(options.FormatUrl(), new GrpcChannelOptions { HttpClient = httpClient });
#else
        return GrpcChannel.ForAddress(options.FormatUrl());
#endif
    }

    /// <summary>
    /// Gets the options set on this service.
    /// </summary>
    public ServiceOptions Options { get; }

    /// <summary>
    /// Gets the gRPC channel used by this service. This channel can be reused
    /// by passing this instance to other service constructors.
    /// </summary>
    protected GrpcChannel Channel { get; }

    /// <summary>
    /// Create call metadata by setting authentication and version headers
    /// </summary>
    /// <param name="request">Protobuf request message to create headers for</param>
    /// <returns></returns>
    protected Task<Metadata> BuildMetadataAsync(IMessage? request = null) {
        var headers = new Metadata() {
            {"TrinsicSDKLanguage", "dotnet"},
            {"TrinsicSDKVersion", GetSdkVersion()}
        };

        // Build authentication header
        if (string.IsNullOrWhiteSpace(Options.AuthToken))
        {
            headers.Add("Authorization", $"Bearer {Options.AuthToken}");
        }
        return Task.FromResult(headers);
    }

    /// <summary>
    /// Create call metadata by setting the required authentication and version headers for provided request
    /// </summary>
    /// <param name="request">Protobuf request message to create headers for</param>
    /// <returns></returns>
    protected Metadata BuildMetadata(IMessage? request = null) {
        var headers = new Metadata() {
            {"TrinsicSDKLanguage", "dotnet"},
            {"TrinsicSDKVersion", GetSdkVersion()}
        };

        // Build authentication header
        if (string.IsNullOrWhiteSpace(Options.AuthToken))
        {
            headers.Add("Authorization", $"Bearer {Options.AuthToken}");
        }
        return headers;
    }

    /// <summary>
    /// Fetches the current version of the SDK
    /// </summary>
    /// <returns></returns>
    protected internal static string GetSdkVersion() {
        // This will always be 1.0.0 on local builds, since the version number is set on the github action during publish.
        return Assembly.GetAssembly(typeof(ServiceBase))?.GetName().Version?.ToString(3) ?? "unknown";
    }
}

File: python/trinsic/service_base.py
"""
Base class and helper methods for the Service wrappers
"""
import base64
from abc import ABC
from typing import Dict

import trinsicokapi.metadata
from betterproto import Message
from grpclib.client import Channel

from trinsicokapi import metadata

import trinsic
from trinsic.proto.sdk.options.v1 import ServiceOptions
from trinsic.proto.services.account.v1 import AccountProfile
from trinsic.proto.services.common.v1 import ResponseStatus
from trinsic.security_providers import (
    OberonSecurityProvider,
    SecurityProvider,
    ITokenProvider,
    MemoryTokenProviderFactory,
)
from trinsic.trinsic_util import trinsic_config, create_channel


class ServiceBase(ABC):
    """
    Base class for service wrapper classes, provides the metadata functionality in a consistent manner.
    """

    def __init__(self, server_config: ServiceOptions, token_provider: ITokenProvider):
        self.service_options: ServiceOptions = server_config or trinsic_config()
        self.token_provider: ITokenProvider = (
            token_provider or MemoryTokenProviderFactory.instance()
        )
        self._channel: Channel = create_channel(self.service_options)
        self._security_provider: SecurityProvider = OberonSecurityProvider()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def __del__(self):
        self.close()

    def close(self):
        """Close the underlying channel"""
        if hasattr(self, "_channel") and self._channel is not None:
            try:
                self._channel.close()
            except RuntimeError:
                # If the event loop is closed, NBD.
                pass

    def set_auth_token(self, auth_token: str) -> None:
        self.service_options.auth_token = auth_token

    def build_metadata(self, request: Message = None) -> Dict[str, str]:
        """
        Create call metadata by setting required authentication headers via `AccountProfile`
        :return: authentication headers with base-64 encoded Oberon
        """
        call_metadata = {
            "TrinsicOkapiVersion".lower(): trinsicokapi.metadata.get_metadata().version,
            "TrinsicSDKLanguage".lower(): "python",
            "TrinsicSDKVersion".lower(): trinsic.__version__(),
        }
        if request is not None:
            auth_token = self.token_provider.get()
            if self.service_options and self.service_options.auth_token:
                auth_token = self.service_options.auth_token

            call_metadata["authorization"] = self._security_provider.get_auth_header(
                AccountProfile().parse(data=base64.urlsafe_b64decode(auth_token)),
                request,
            )
        return call_metadata

    @property
    def channel(self):
        """Underlying channel"""
        return self._channel


class ResponseStatusException(Exception):
    """
    Exception raised when an action has a non-success response status.
    """

    def __init__(self, action: str, status: ResponseStatus):
        super().__init__(f"{action}, status={repr(status)}")
        self.status = status

    @staticmethod
    def assert_success(status: ResponseStatus, action: str) -> None:
        if status != ResponseStatus.SUCCESS:
            raise ResponseStatusException(action, status)
